cfep-09.md

```diff
@@ -0,0 +1,65 @@
+
+<table>
+<tr><td> Pinning Proposal System for Automatic Rebuilds</td><td> Pinning Proposal </td>
+<tr><td> Status </td><td> Proposed </td></tr>
+<tr><td> Author(s) </td><td> Christopher J. "CJ" Wright &lt;cjwright4242@gmail.com&gt;</td></tr>
+<tr><td> Created </td><td> Feb 20, 2019</td></tr>
+<tr><td> Updated </td><td> Aug 30, 2016</td></tr>
+<tr><td> Discussion </td><td> NA </td></tr>
+<tr><td> Implementation </td><td> NA </td></tr>
+</table>
+
+## Abstract
+
+This CFEP proposes a change to how our internal pinning system works.
+The new system will comprise of a new repo for proposed migrations.
+To propose a migration a PR will be issued with a new version of
+the global pinning file and some metadata about the migration needed 
+(if any).
+Once the PR is squashed and merged the bot will take the metadata
+to calculate which packages need to be rebuilt and start issuing
+PRs with the new pinning as a local pinning file to the feedstocks
+in topo order.
+
+This CFEP is taken from @beckermr's work on [conda-forge.github.io](https://github.com/conda-forge/conda-forge.github.io/issues/712).
+
+## Introduction
+Currently when we move pinnings we move the pin in the ``conda-forge-pinning-feedstock``.
+Then we start a migration using the ``regro-cf-autotick-bot``.
+This order of operations causes some issues, when new feedstocks
+are created they recieve the new pinnings even if their dependencies
+are not rebuilt with the new pinnings themselves.
+
+## Specification
+To change the pinnings a PR will be opened into the ``migration`` directory in a proposed pinning repo.
+The PR will contain a complete copy of the global pinning file with any changes
+needed for the migration and a metadata file which includes the name of the migration,
+the source nodes which will be changed, and the finishing criteria (eg. 80% finished and 
+``numpy`` must be migrated).
+Once the commits are squashed and merged the bot will read the new file and start
+issuing PRs in topo order which bump the build number, put a copy of the modified
+pinning file into the repo for conda-smithy/build to find and rerendering the feedstock.
```

Comment:
This will override the global pinning file and then we will have different pinnings scattered in feedstocks which will not be updated using a conda-smithy rerender. (OTOH, current .ci_support./*.yaml files are updated using conda-smithy rerender)

Reply:
We could have rerender clear out the file once the migration has been merged. I guess we want new pinning without a topo rebuild if the ABI hasn't broken.

Reply:
It might be easier to handle a yaml snippet, since it would only override the bit of the pinning that it needs to leaving the rest to the global pinning, but then we need to use something like [``ChainDB``](https://github.com/xonsh/xonsh/blob/91438e56b09947f69bdd938a9ccf27da59f60a43/xonsh/lib/collections.py) to perform the dict merging inside of conda-build.

Reply:
How about storing these in a separate file (like `.ci_support/new_pinnings.yaml`) and make conda-smithy use the values from them only if they are newer than what is in the global pinnings file? (This is similar to what we did for py37 in conda-forge.yml file)

Reply:
So the general idea is to have a few of these that get fused at smithy time into the effective pinning file and then variant configs get generated as normal. That fusing process also deals with expiration of completed migrations etc.

Reply:
We want to have these be named somewhat arbitrariy of these as we can very likely have a number of migrations in flight simultaneously.

The other thing we want to avoid is needing to have to change both smithy and pinnings for every migration we run

Reply:
I still don't understand what changes are needed in a feedstock

Reply:
See my comment below.